packages/flutter/lib/src/rendering/slivers.dart

```diff
@@ -0,0 +1,1850 @@
+  ///
+  /// For example, if the constraints have a scroll offset of 100 and a
+  /// remaining paint extent of 100, and the arguments to this method describe
+  /// the region 50..150, then the returned value would be 50 (from scroll
+  /// offset 100 to scroll offset 150).
+  ///
+  /// This method is not useful if there is not a 1:1 relationship between
+  /// consumed scroll offset and consumed paint extent. For example, if the
+  /// sliver always paints the same amount but consumes a scroll offset extent
+  /// that is proportional to the [SliverConstraints.scrollOffset], then this
+  /// function's results will not be consistent.
+  double calculatePaintOffset(SliverConstraints constraints, { @required double from, @required double to }) {
+    assert(from <= to);
+    double a = constraints.scrollOffset;
+    double b = constraints.scrollOffset + constraints.remainingPaintExtent;
+    // the clamp on the next line is to avoid floating point rounding errors
+    return (to.clamp(a, b) - from.clamp(a, b)).clamp(0.0, constraints.remainingPaintExtent);
+  }
+
+  /// Returns the distance from the leading _visible_ edge of the sliver to the
+  /// side of the given child closest to that edge.
+  ///
+  /// For example, if the [constraints] describe this sliver as having an axis
+  /// direction of [AxisDirection.down], then this is the distance from the top
+  /// of the visible portion of the sliver to the top of the child. On the other
+  /// hand, if the [constraints] describe this sliver as having an axis
+  /// direction of [AxisDirection.up], then this is the distance from the bottom
+  /// of the visible portion of the sliver to the bottom of the child. In both
+  /// cases, this is the direction of increasing
+  /// [SliverConstraints.scrollOffset] and
+  /// [SliverLogicalParentData.scrollOffset].
+  ///
+  /// Calling this for a child that is not visible is not valid.
+  ///
+  /// For children that are [RenderSliver]s, the leading edge of the _child_
+  /// will be the leading _visible_ edge of the child, not the part of the child
+  /// that would locally be a scroll offset 0.0. For children that are not
+  /// [RenderSliver]s, for example a [RenderBox] child, it's the actual distance
+  /// to the edge of the box, since those boxes do not know how to handle being
+  /// scrolled.
+  ///
+  /// This is used by [RenderSliverHelpers.hitTestBoxChild]. If you do not use
+  /// the [RenderSliverHelpers] mixin and do not call this method yourself, you
+  /// do not need to implement this method.
+  @protected
+  double childPosition(@checked RenderObject child) {
+    assert(() {
+      throw new FlutterError('$runtimeType does not implement childPosition.');
+    });
+    return 0.0;
+  }
+
+  @override
+  void applyPaintTransform(RenderObject child, Matrix4 transform) {
+    assert(() {
+      throw new FlutterError('$runtimeType does not implement applyPaintTransform.');
+    });
```

Comment:
The base class has a default implementation which makes sense for most kinds of RenderObject.
For RenderSliver, there's no reasonable default paint transform.
What I'd really like to do is "re-abstract" the method, but Dart doesn't support that yet (see https://github.com/dart-lang/sdk/issues/28250 for a suggested `@mustOverride` which we could use though).